source/common/event/scaled_range_timer.cc

```diff
@@ -0,0 +1,203 @@
+#include "common/event/scaled_range_timer.h"
+
+#include <chrono>
+#include <cmath>
+#include <memory>
+
+#include "envoy/event/timer.h"
+
+#include "common/common/assert.h"
+#include "common/common/scope_tracker.h"
+
+namespace Envoy {
+namespace Event {
+
+/**
+ * Implementation of RangeTimer that can be scaled by the backing manager object.
+ *
+ * Instances of this class exist in one of 4 states:
+ *  - inactive: not enabled
+ *  - pending: enabled, min timeout not elapsed
+ *  - active: enabled, min timeout elapsed, max timeout not elapsed
+ */
+class ScaledRangeTimerManager::ScaledRangeTimer final : public RangeTimer {
+public:
+  ScaledRangeTimer(TimerCb callback, ScaledRangeTimerManager& manager)
+      : manager_(manager), callback_(callback),
+        pending_timer_(manager.dispatcher_.createTimer([this] { onPendingTimerComplete(); })) {}
+
+  ~ScaledRangeTimer() override { disableTimer(); }
+
+  void disableTimer() override {
+    struct Dispatch {
+      Dispatch(ScaledRangeTimer& timer) : timer(timer) {}
+      ScaledRangeTimer& timer;
+      void operator()(const Inactive&) {}
+      void operator()(const Pending&) { timer.pending_timer_->disableTimer(); }
+      void operator()(const Active& active) {
+        timer.manager_.disableActiveTimer(active.duration, active.bucket_position);
+      }
+    };
+    absl::visit(Dispatch(*this), state_);
+    state_.emplace<Inactive>();
+  }
+
+  void enableTimer(const std::chrono::milliseconds& min_ms, const std::chrono::milliseconds& max_ms,
+                   const ScopeTrackedObject* scope) override {
+    disableTimer();
+    const MonotonicTime now = manager_.dispatcher_.timeSource().monotonicTime();
+    const MonotonicTime latest_trigger = now + max_ms;
+    scope_ = scope;
+    ENVOY_LOG_MISC(info, "enableTimer called on {} for ({}ms, {}ms)", static_cast<void*>(this),
+                   min_ms.count(), max_ms.count());
+    if (min_ms <= std::chrono::milliseconds::zero()) {
+      auto it = manager_.add(*this, max_ms);
+      state_.emplace<Active>(max_ms, it);
+    } else {
+      state_.emplace<Pending>(latest_trigger);
+      pending_timer_->enableTimer(min_ms);
+    }
+  }
+
+  bool enabled() override { return !absl::holds_alternative<Inactive>(state_); }
+
+  void trigger() {
+    ASSERT(absl::holds_alternative<Active>(state_));
+    state_.emplace<Inactive>();
+    if (scope_ == nullptr) {
+      callback_();
+    } else {
+      ScopeTrackerScopeState scope(scope_, manager_.dispatcher_);
+      callback_();
+    }
+    scope_ = nullptr;
```

Comment:
there is potential for use after free by writing to scope_ after execution of callback_.  Also, re-enabling the timer with a scope from within the callback will result in scope_ being cleared after the callback executes.

Consider doing this instead:
    } else {
      ScopeTrackerScopeState scope(scope_, manager_.dispatcher_);
      scope_ = nullptr;
      callback_();
    }

Please write a test that covers deletion and reset of the RangeTimer from the callback.